File: src/main/java/org/funivan/lologa/tile/Position/Next.java
package org.funivan.lologa.tile.Position;

import org.funivan.lologa.tile.TileInterface;

public class Next implements PositionInterface {
    private final PositionInterface start;
    private final int rowStep;
    private final int colStep;

    public Next(PositionInterface start, int rowStep, int colStep) {
        this.start = start;
        this.rowStep = rowStep;
        this.colStep = colStep;
    }

    public Next(TileInterface start, int rowStep, int colStep) {
        this(start.position(), rowStep, colStep);
    }

    @Override
    public final int row() {
        return this.start.row() + this.rowStep;
    }

    @Override
    public final int col() {
        return this.start.col() + this.colStep;
    }

    @Override
    public boolean equals(Object o) {
        boolean result = false;
        if (this == o) {
            result = true;
        } else if (o instanceof PositionInterface) {
            result = (((PositionInterface) o).row() == this.row() && ((PositionInterface) o).col() == this.col());
        }
        return result;
    }

    @Override
    public int hashCode() {
        return 31 * this.row() + this.col();
    }


    public static class Top extends Next {
        public Top(TileInterface start) {
            super(start, -1, 0);
        }
    }

    public static class Right extends Next {
        public Right(TileInterface start) {
            super(start, 0, 1);
        }
    }

    public static class Bottom extends Next {
        public Bottom(TileInterface start) {
            super(start, 1, 0);
        }
        public Bottom(PositionInterface start) {
            super(start, 1, 0);
        }
    }

    public static class Left extends Next {
        public Left(TileInterface start) {
            super(start, 0, -1);
        }
    }
}

File: src/main/java/org/funivan/lologa/tile/Position/Next.kt
package org.funivan.lologa.tile.Position

import org.funivan.lologa.tile.TileInterface

open class Next(private val start: PositionInterface, private val rowStep: Int, private val colStep: Int) : PositionInterface {

    constructor(start: TileInterface, rowStep: Int, colStep: Int) : this(start.position(), rowStep, colStep) {}

    override fun row(): Int {
        return this.start.row() + this.rowStep
    }

    override fun col(): Int {
        return this.start.col() + this.colStep
    }

    override fun equals(o: Any?): Boolean {
        var result = false
        if (this === o) {
            result = true
        } else if (o is PositionInterface) {
            result = o.row() == this.row() && o.col() == this.col()
        }
        return result
    }

    override fun hashCode(): Int {
        return 31 * this.row() + this.col()
    }


    class Top(start: TileInterface) : Next(start, -1, 0)

    class Right(start: TileInterface) : Next(start, 0, 1)

    class Bottom : Next {
        constructor(start: TileInterface) : super(start, 1, 0) {}
        constructor(start: PositionInterface) : super(start, 1, 0) {}
    }

    class Left(start: TileInterface) : Next(start, 0, -1)
}
